### PR TITLE
Update certificate to Pending instead of Failed when issuer is not ready

### PR DESCRIPTION
**Pull request by ARichman555**
### Issue # (if applicable)

N/A

### Reason for this change

This will allow cert-manager to recover itself after the pod has downtime. As of now, the certificate requests will enter a terminal failed state which means cert-manager will never retry. By putting these in Pending, we can retry according to cert-manager's retry strategy. 

### Description of changes
* Updates the certificate status to Pending instead of Failed if the issuer is not found or is not ready
* Moves the provisioner to always be made at issuance time (prevents branching logic)
* Re-creates the provisioner if it does not already exist at issuance time
* Adds unit tests for everything

### Describe any new or updated permissions being added

There are no new permissions being updated or added.

### Description of how you validated changes

Added unit tests and tested locally with e2e tests. Also manually tested edge cases presented by customer

### CODE REVIEW

File: pkg/aws/.pca.go.swn
Comment (bmsiegel) [inline pkg/aws/.pca.go.swn:1]: What is this

Comment (ARichman555) [reply pkg/aws/.pca.go.swn:1]: Vim swap file, removed

File: pkg/controllers/certificaterequest_controller.go [hunk 1]
```diff
@@ -145,22 +150,26 @@ func (r *CertificateRequestReconciler) Reconcile(ctx context.Context, req ctrl.R
...
+	if !exists {
+		config, err := awspca.GetConfig(ctx, r.Client, iss.GetSpec())
+		if err != nil {
+			log.Error(err, "error loading config")
+			_ = r.setStatus(ctx, cr, cmmeta.ConditionFalse, cmapi.CertificateRequestReasonFailed, "failed to retrieve aws config")
+			return ctrl.Result{}, err
+		}
+
+		provisioner = NewProvisioner(config, iss.GetSpec().Arn, issuerName)
 	}
```
Comment (anbaig) [inline pkg/controllers/certificaterequest_controller.go:173]: Can we just hide all this logic inside of `GetProvisioner`. From the perspective of the `CertificateRequestController` it really only cares that GetProvisoner returns something -- not how.

File: pkg/aws/pca.go [hunk 2]
```diff
@@ -79,21 +157,24 @@ func StoreProvisioner(name types.NamespacedName, provisioner GenericProvisioner)
...
 
 // idempotencyToken is limited to 64 ASCII characters, so make a fixed length hash.
 // @see: https://docs.aws.amazon.com/AWSEC2/latest/APIReference/Run_Instance_Idempotency.html
 func idempotencyToken(cr *cmapi.CertificateRequest) string {
-    token := []byte(cr.ObjectMeta.Namespace + "/" + cr.ObjectMeta.Name)
-    fullHash := fmt.Sprintf("%x", sha256.Sum256(token))
-    return fullHash[:36] // Truncate to 36 characters
+	token := []byte(cr.ObjectMeta.Namespace + "/" + cr.ObjectMeta.Name)
+	fullHash := fmt.Sprintf("%x", sha256.Sum256(token))
+	return fullHash[:36] // Truncate to 36 characters
```
Comment (anbaig) [inline pkg/aws/pca.go:177]: Is there a way to break all these stylistic changes into a separate PR? It's a bit distracting here.

File: pkg/aws/pca.go [hunk 3]
```diff
@@ -62,15 +74,81 @@ type PCAProvisioner struct {
...
+			config.WithRegion(spec.Region),
+		)
+	}
+
+	return config.LoadDefaultConfig(ctx)
+}
+
+func ClearProvisioners() {
+	collection.Clear()
+}
```
Comment (bmsiegel) [inline pkg/aws/pca.go:142]: Oof it's a testing function in service code?

File: pkg/aws/pca.go [hunk 4]
```diff
@@ -62,15 +74,81 @@ type PCAProvisioner struct {
...
 // GetProvisioner gets a provisioner that has previously been stored
 func GetProvisioner(name types.NamespacedName) (GenericProvisioner, bool) {
 	value, exists := collection.Load(name)
 	if !exists {
 		return nil, exists
 	}
 
-	p, exists := value.(GenericProvisioner)
-	return p, exists
+	return value.(GenericProvisioner), exists
```
Comment (bmsiegel) [inline pkg/aws/pca.go:151]: I don't really get how this is equivalent to above tbh

Comment (ARichman555) [reply pkg/aws/pca.go:151]: Hmmm my thinking was that we can always assume the map has a GenericProvisioner, so if it's not then we should panic since that means something's gone quite wrong. But thinking about it now, having it re-create the provisioner is probably better. I'll revert this

File: pkg/aws/pca.go (hunk 3 above)
Comment (ARichman555) [reply pkg/aws/pca.go:142]: Yea figured it's better to leave the provisioner map as private and add this than make the provisioner map public. As far as I can tell Go doesn't really have test only functions :(. But I think this should be ok to just do since we only call it in the tests. If not, we can just use different namespaces for the tests that need different provisioners.

File: pkg/aws/pca.go (hunk 2 above)
Comment (ARichman555) [reply pkg/aws/pca.go:177]: this is auto added by the linter. Some of the other renaming is necessary because I'm adding conflicting imports. I'll try to look through and cut out anything that's not necessary

File: pkg/aws/pca.go (hunk 3 above)
Comment (bmsiegel) [reply pkg/aws/pca.go:142]: I guess this isn't really super dangerous to call, it just means we'll re-create it

File: pkg/aws/pca.go (hunk 4 above)
Comment (bmsiegel) [reply pkg/aws/pca.go:151]: I just don't understand how the exists is populated, I'm ok leaving it if it makes sense

File: pkg/controllers/certificaterequest_controller.go (hunk 1 above)
Comment (anbaig) [reply pkg/controllers/certificaterequest_controller.go:173]: I think the only appropriate part to leave here is the:

```
		if err != nil {
			log.Error(err, "error loading config")
			_ = r.setStatus(ctx, cr, cmmeta.ConditionFalse, cmapi.CertificateRequestReasonFailed, "failed to retrieve aws config")
			return ctrl.Result{}, err
		}
```

which should get run if `GetProvisoner` returns an error.

File: pkg/aws/pca.go [hunk 5]
```diff
@@ -62,38 +74,103 @@ type PCAProvisioner struct {
...
+	collection.Clear()
 }
 
-// NewProvisioner returns a new PCAProvisioner
-func NewProvisioner(config aws.Config, arn string) (p *PCAProvisioner) {
-	return &PCAProvisioner{
+// GetProvisioner gets a provisioner that has previously been stored or creates a new one
+func GetProvisioner(ctx context.Context, client client.Client, name types.NamespacedName, spec *api.AWSPCAIssuerSpec) (GenericProvisioner, error) {
+	value, _ := collection.Load(name)
+	p, isProvisioner := value.(GenericProvisioner)
```
Comment (bmsiegel) [inline pkg/aws/pca.go:147]: Oh nice, this worked!

Comment (bmsiegel) [reply pkg/aws/pca.go:147]: Is there any way to add a unit test to check that this works?

Comment (ARichman555) [reply pkg/aws/pca.go:147]: We unit test this in the certificate_request_controller_test. All of those start with an empty collection so they should all hit this code path. We also test the caching in a separate unit test in that file

Comment (ARichman555) [reply pkg/aws/pca.go:147]: Also saw online that the type assertion first checks that it's non-null and then asserts the type

Comment (bmsiegel) [reply pkg/aws/pca.go:147]: Nicem sounds good

File: pkg/controllers/certificaterequest_controller_test.go [hunk 6]
```diff
@@ -60,14 +60,47 @@ func (p *fakeProvisioner) Get(ctx context.Context, cr *cmapi.CertificateRequest,
 	return p.cert, p.caCert, p.getErr
 }
 
-type createMockProvisioner func()
-
 func TestProvisonerOperation(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ARichman555) [reply pkg/controllers/certificaterequest_controller_test.go]: moved to pca_test.go

File: e2e/k8_helpers.go [hunk 7]
```diff
@@ -37,7 +37,7 @@ func waitForIssuerReady(ctx context.Context, client *clientV1beta1.Client, name
 }
 
 func waitForClusterIssuerReady(ctx context.Context, client *clientV1beta1.Client, name string) error {
-	return wait.PollImmediate(250*time.Millisecond, 2*time.Minute,
+	return wait.PollImmediate(250*time.Millisecond, 3*time.Minute,
```
Comment (anbaig) [inline e2e/k8_helpers.go:40]: Why is this changed?

File: e2e/k8_helpers.go [hunk 8]
```diff
@@ -56,6 +56,39 @@ func waitForClusterIssuerReady(ctx context.Context, client *clientV1beta1.Client
 		})
 }
 
+func waitForCertificateRequestPending(ctx context.Context, client *cmclientv1.CertmanagerV1Client, name string, namespace string) error {
+	return wait.PollImmediate(250*time.Millisecond, 2*time.Minute,
```
Comment (anbaig) [inline e2e/k8_helpers.go:60]: any reason for these timing values?

File: pkg/aws/.pca.go.swn
[the remark this answers is not in the document]
Comment (ARichman555) [reply pkg/aws/.pca.go.swn:1]: removed again

File: e2e/e2e_test.go [hunk 9]
```diff
@@ -551,3 +552,100 @@ func TestNamespaceIssuers(t *testing.T) {
 		}
 	}
 }
+
+func TestCertificateRecovery(t *testing.T) {
```
Comment (anbaig) [inline e2e/e2e_test.go:556]: Can we make this test name long, descriptive, and possibly ugly? I want it to be ultra-clear what this test is validating.

File: pkg/aws/.pca.go.swn
[the remark this answers is not in the document]
Comment (anbaig) [reply pkg/aws/.pca.go.swn:1]: Put it in `.gitignore` and it will never bother us again :)

File: e2e/k8_helpers.go (hunk 8 above)
Comment (ARichman555) [reply e2e/k8_helpers.go:60]: I just based it on the other functions

File: e2e/k8_helpers.go (hunk 7 above)
Comment (ARichman555) [reply e2e/k8_helpers.go:40]: ah I was messing with things, will revert
